### PR TITLE
[feat: CL Hooks]: Implement hooks for concentrated liquidity operations

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #XXX

## What is the purpose of the change

> Note to reviewer: the relevant diff in this PR is in `pool_hooks.go` and `pool_hooks_test.go`. The rest is mostly wiring and source code for the reference CW contract used in testing. **This PR is state compatible.**

This PR implements the core abstractions needed for CL hooks. It is added in the CL module instead of pool manager because we expect to want hooks to be exposed for many CL operations, not just swaps. It might make sense in the future to abstract this logic out even further into osmoutils given how there are overlapping design patterns between CL hooks and before send hooks implemented in tokenfactory.

**This PR is state compatible** and simply implements and tests the necessary abstractions to be able to add arbitrary types of CL hooks to various CL functions. There was a meaningful amount of testing work that required implementing a reference CW contract, which should hopefully simplify testing higher level functions related to CL hooks.

## Testing and Verifying

- All new logic should be tested in `pool_hooks_test.go`

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/types/constants.go [hunk 1]
```diff
@@ -28,6 +28,16 @@ const (
...
+	// This is a (very generous) gas limit intended to protect against CL hooks that are
+	// executed with malicious intent in begin block code.
+	//
+	// This is an unlikely scenario as long as contract deployment is gated by governance
+	// and protorev/txfee swaps are on whitelisted tokens, but letting contract calls take
+	// unbounded gas in these scenarios is a risk we don't want to take regardless.
+	//
+	// 2M gas is enough to execute tens of expensive CL operations and is only set this high
+	// to accommodate position withdrawals, which are unusually expesive.
+	ContractHookGasLimit = 2_000_000
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/types/constants.go:40]: Note: this is a tentative value that might be a bit too generous. Since position withdrawals are extremely expensive (700k+ gas) it's hard to go too low here, but we should still choose a value that would not be exploitable post-permissionless CW contracts

File: x/concentrated-liquidity/testcontracts/contract-sources/counter/.cargo/config [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[alias]
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/testcontracts/contract-sources/counter/.cargo/config:1]: Note: most of the remaining diff past this point is source code for the CW contract used for testing. I only included it here as a reference and to ensure tests can be extended in the future by other contributors (black-box wasm is hard to work with)

File: x/concentrated-liquidity/pool_hooks.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		ctx.GasMeter().ConsumeGas(childCtx.GasMeter().GasConsumed(), "Track CL action contract call gas")
+	}
+
+	return nil
+}
+
+// --- Store helpers ---
+
+// GetPoolHookPrefixStore returns the substore for a specific pool ID where hook-related data is stored.
+func (k Keeper) GetPoolHookPrefixStore(ctx sdk.Context, poolID uint64) sdk.KVStore {
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks.go:56]: Do we need to have this exported?

File: x/concentrated-liquidity/types/errors.go [hunk 4]
```diff
@@ -913,3 +913,11 @@ type LastPositionTransferError struct {
 func (e LastPositionTransferError) Error() string {
 	return fmt.Sprintf("cannot transfer position %d because it is the last position in pool %d", e.PositionId, e.PoolId)
 }
+
+type ContractHookOutOfGasError struct {
+	GasLimit uint64
+}
+
+func (e ContractHookOutOfGasError) Error() string {
+	return fmt.Sprintf("A single contract call cannot exceed %d gas in a CL hook call.", e.GasLimit)
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/errors.go:922]: Should we also format how much gas it actually took?

File: x/concentrated-liquidity/types/constants.go [hunk 5]
```diff
@@ -28,6 +28,16 @@ const (
...
 	BaseGasFeeForTransferPosition       = 10_000
+	// This is a (very generous) gas limit intended to protect against CL hooks that are
+	// executed with malicious intent in begin block code.
+	//
+	// This is an unlikely scenario as long as contract deployment is gated by governance
+	// and protorev/txfee swaps are on whitelisted tokens, but letting contract calls take
+	// unbounded gas in these scenarios is a risk we don't want to take regardless.
+	//
+	// 2M gas is enough to execute tens of expensive CL operations and is only set this high
+	// to accommodate position withdrawals, which are unusually expesive.
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/constants.go:39]:
```suggestion
	// to accommodate position withdrawals, which are unusually expensive.
```

File: x/concentrated-liquidity/types/constants.go (hunk 1 above)
Comment (p0mvn) [reply x/concentrated-liquidity/types/constants.go:40]: Given concerns, can we make this a module param so that we can tweak this in the future without a major upgrade?

File: x/concentrated-liquidity/pool_hooks_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+			s.Require().NoError(err)
+
+			// --- System under test ---
+
+			err = s.clk.CallPoolActionListener(s.Ctx, msgBz, validPoolId, validActionPrefix)
+
+			// --- Assertions ---
+
+			if tc.expectedError != nil {
+				s.Require().ErrorAs(err, &tc.expectedError)
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks_test.go:161]: Why not `ErrorIs`? `ErrorAs` only checks error type but not the specific message and fields

File: x/concentrated-liquidity/pool_hooks_test.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+// that counts `Amount` times and does a state write on each iteration.
+type CountMsg struct {
+	Amount int64 `json:"amount"`
+}
+type CountSudoMsg struct {
+	Count CountMsg `json:"count"`
+}
+
+// TestSetAndGetPoolHookContract tests the basic functionality of setting and getting a CW contract address for a specific hook type from state.
+func (s *KeeperTestSuite) TestSetAndGetPoolHookContract() {
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks_test.go:30]: Please test deletion

File: x/concentrated-liquidity/pool_hooks.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+// nolint: unused
+// setPoolHookContract sets the contract address linked to the passed in hook for a specific pool ID.
+func (k Keeper) setPoolHookContract(ctx sdk.Context, poolID uint64, actionPrefix string, cosmwasmAddress string) error {
+	store := k.GetPoolHookPrefixStore(ctx, poolID)
+
+	// If cosmwasm address is nil, treat this as a delete operation for the stored address.
+	if cosmwasmAddress == "" {
+		store.Delete([]byte(actionPrefix))
+		return nil
+	}
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks.go:84]: nit: consider adding a separate deletion helper to keep this setter abstraction clean

File: x/concentrated-liquidity/pool_hooks.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	bz := store.Get([]byte(actionPrefix))
+	if bz == nil {
+		return ""
+	}
+
+	return string(bz)
+}
+
+// nolint: unused
+// setPoolHookContract sets the contract address linked to the passed in hook for a specific pool ID.
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks.go:76]: If we continue deleting on the empty address, let's make the godoc more detailed

File: x/concentrated-liquidity/pool_hooks.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	}()
+
+	cosmwasmAddress := k.GetPoolHookContract(ctx, poolId, actionPrefix)
+	if cosmwasmAddress != "" {
+		cwAddr, err := sdk.AccAddressFromBech32(cosmwasmAddress)
+		if err != nil {
+			return err
+		}
+
+		em := sdk.NewEventManager()
```
Comment (p0mvn) [inline x/concentrated-liquidity/pool_hooks.go:32]: Why do we create a new event manager?

AFAIU, this leads to contract events being dropped. Why is that the goal?

**Review by p0mvn — COMMENTED** [review]
Nice work. Only had nits and some questions.

Should be a quick re-review once questions are answered

File: x/concentrated-liquidity/testcontracts/contract-sources/counter/contracts/counter/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
+use cw2::set_contract_version;
+
+use crate::error::ContractError;
+use crate::msg::{InstantiateMsg, SudoMsg};
+use crate::state::{State, STATE};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:infinite-track-beforesend";
```
Comment (mattverse) [inline x/concentrated-liquidity/testcontracts/contract-sources/counter/contracts/counter/src/contract.rs:11]:
```suggestion
const CONTRACT_NAME: &str = "crates.io:counter";
```

**Review by mattverse — COMMENTED** [review]
LGTM, @p0mvn's comments seem to be addressing what needs to be resolved before merge!

One question for me before a ACK for this PR would be understanding where `callPoolActionListener` and `setPoolHookContract` would be used in the future.

File: x/concentrated-liquidity/pool_hooks.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+// If no contract is linked to the hook, this function is a no-op.
+func (k Keeper) callPoolActionListener(ctx sdk.Context, msgBz []byte, poolId uint64, actionPrefix string) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = types.ContractHookOutOfGasError{GasLimit: types.ContractHookGasLimit}
+		}
+	}()
+
+	cosmwasmAddress := k.GetPoolHookContract(ctx, poolId, actionPrefix)
+	if cosmwasmAddress != "" {
```
Comment (PaddyMc) [inline x/concentrated-liquidity/pool_hooks.go:26]: This PR looks great! I only have small nit to do with style. 

```suggestion
	if cosmwasmAddress == "" {
	    return nil
	}
	...
```

I find it's usually best to return early and continue execution than to indent with statement, but this isn't a big issue so feel free to ignore :smile:

File: x/concentrated-liquidity/types/errors.go (hunk 4 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/types/errors.go:922]: Hmm I don't think this is possible since we would exit once the limit is hit (so the gas consumed would be the limit amount).

Since the attack vector here is unbounded gas consumption with an infinite loop, it wouldn't make sense to continue metering until execution completes so we know how much to put in a field like this.

File: x/concentrated-liquidity/pool_hooks.go (hunk 12 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks.go:26]: Great suggestion, thanks! 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/testcontracts/contract-sources/counter/contracts/counter/src/contract.rs (hunk 11 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/testcontracts/contract-sources/counter/contracts/counter/src/contract.rs:11]: 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks.go]: Good point, wasn’t necessary. Also unexported `GetPoolHookContract` 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/types/constants.go (hunk 5 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/types/constants.go]: Fixed 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks_test.go (hunk 6 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks_test.go]: Thanks for flagging. This was from an earlier version of the error message that tried to include more information like actual gas consumed (see response to your other comment on this for why that isn’t easily doable). Changed to `ErrorIs` here. 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks.go (hunk 8 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks.go]: Might be slightly overkill but I agree with keeping set/delete abstractions separate so added here 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks.go (hunk 9 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks.go:78]: Expanded godoc 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks.go (hunk 10 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks.go:32]: Hmm so for this one, I was following juno’s endblock hook design pattern. I'm not sure what exactly the purpose is here, but I assumed it might have something to do with an event spam vector.

Unless you feel it's important enough to not drop events here for us to dig deeper into this tradeoff space/attack surface, I'd vote for keeping it as is, but open to exploring more if you feel it is important to change

File: x/concentrated-liquidity/types/constants.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/types/constants.go]: Great point. Made this change here: 46aa7f26684c44b455130981a18944005b331e6c

File: x/concentrated-liquidity/pool_hooks_test.go (hunk 7 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/pool_hooks_test.go:30]: Added tests for these as well and related edge case (e.g. attempting to delete when no contract was set)

File: x/concentrated-liquidity/pool_hooks.go (hunk 10 above)
Comment (p0mvn) [reply x/concentrated-liquidity/pool_hooks.go]: I don't have strong opinions since we can change this in the next upgrade if needed. I'm just thinking that events might be useful for debugging but we are currently dropping them

File: x/concentrated-liquidity/types/params.go [hunk 13]
```diff
@@ -95,6 +101,7 @@ func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
 		paramtypes.NewParamSetPair(KeyDiscountRate, &p.BalancerSharesRewardDiscount, validateBalancerSharesDiscount),
 		paramtypes.NewParamSetPair(KeyAuthorizedUptimes, &p.AuthorizedUptimes, validateAuthorizedUptimes),
 		paramtypes.NewParamSetPair(KeyUnrestrictedPoolCreatorWhitelist, &p.UnrestrictedPoolCreatorWhitelist, osmoutils.ValidateAddressList),
+		paramtypes.NewParamSetPair(KeyHookGasLimit, &p.HookGasLimit, validateHookGasLimit),
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/params.go:104]: Let's make sure that we set this param in the upgrade handler. I suspect that this is what might be breaking e2e

**Review by p0mvn — COMMENTED** [review]
LGTM once e2e is passing. Please lmk once it does for approval

**Review by p0mvn — APPROVED** [review]
LGTM
